File: mobile/lib/features/groups/models/group_permission.dart
import 'group_role.dart';

class GroupPermission {
  final String value;
  final String label;
  final String description;

  const GroupPermission({
    required this.value,
    required this.label,
    required this.description,
  });

  factory GroupPermission.fromJson(Map<String, dynamic> json) {
    return GroupPermission(
      value: json['value'] as String,
      label: json['label'] as String,
      description: json['description'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'value': value,
      'label': label,
      'description': description,
    };
  }

  // Common permission constants
  static const String editGroup = 'edit_group';
  static const String deleteGroup = 'delete_group';
  static const String changeGroupSettings = 'change_group_settings';
  static const String uploadGroupAvatar = 'upload_group_avatar';

  static const String addMembers = 'add_members';
  static const String removeMembers = 'remove_members';
  static const String changeMemberRoles = 'change_member_roles';
  static const String viewMemberDetails = 'view_member_details';
  static const String approveMemberRequests = 'approve_member_requests';

  static const String createSessions = 'create_sessions';
  static const String editSessions = 'edit_sessions';
  static const String deleteSessions = 'delete_sessions';
  static const String manageAttendance = 'manage_attendance';

  static const String createPaymentRequests = 'create_payment_requests';
  static const String managePayments = 'manage_payments';
  static const String viewFinancialReports = 'view_financial_reports';

  static const String sendNotifications = 'send_notifications';
  static const String moderateDiscussions = 'moderate_discussions';

  static const String viewGroup = 'view_group';
  static const String joinSessions = 'join_sessions';
  static const String leaveGroup = 'leave_group';
  static const String viewBasicStats = 'view_basic_stats';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is GroupPermission &&
          runtimeType == other.runtimeType &&
          value == other.value;

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => 'GroupPermission(value: $value, label: $label)';
}

class UserPermissions {
  final String? role;
  final String? roleName;
  final List<GroupPermission> permissions;

  const UserPermissions({
    required this.role,
    required this.roleName,
    required this.permissions,
  });

  factory UserPermissions.fromJson(Map<String, dynamic> json) {
    return UserPermissions(
      role: json['role'] as String?,
      roleName: json['role_name'] as String?,
      permissions: (json['permissions'] as List<dynamic>?)
              ?.map((p) => GroupPermission.fromJson(p as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  /// Check if user has a specific permission
  bool hasPermission(String permission) {
    return permissions.any((p) => p.value == permission);
  }

  /// Check if user has any of the specified permissions
  bool hasAnyPermission(List<String> permissionList) {
    return permissionList.any((permission) => hasPermission(permission));
  }

  /// Check if user has all of the specified permissions
  bool hasAllPermissions(List<String> permissionList) {
    return permissionList.every((permission) => hasPermission(permission));
  }

  /// Get user's role as enum
  GroupRole? get roleEnum => GroupRole.fromValue(role);

  Map<String, dynamic> toJson() {
    return {
      'role': role,
      'role_name': roleName,
      'permissions': permissions.map((p) => p.toJson()).toList(),
    };
  }

  @override
  String toString() =>
      'UserPermissions(role: $role, roleName: $roleName, permissions: ${permissions.length})';
}

File: api/app/Enums/GroupPermission.php
<?php

namespace App\Enums;

enum GroupPermission: string
{
    // Group Management
    case EDIT_GROUP = 'edit_group';
    case DELETE_GROUP = 'delete_group';
    case CHANGE_GROUP_SETTINGS = 'change_group_settings';
    case UPLOAD_GROUP_AVATAR = 'upload_group_avatar';
    
    // Member Management
    case ADD_MEMBERS = 'add_members';
    case REMOVE_MEMBERS = 'remove_members';
    case CHANGE_MEMBER_ROLES = 'change_member_roles';
    case VIEW_MEMBER_DETAILS = 'view_member_details';
    case APPROVE_MEMBER_REQUESTS = 'approve_member_requests';
    
    // Content Management
    case CREATE_SESSIONS = 'create_sessions';
    case EDIT_SESSIONS = 'edit_sessions';
    case DELETE_SESSIONS = 'delete_sessions';
    case MANAGE_ATTENDANCE = 'manage_attendance';
    
    // Financial Management
    case CREATE_PAYMENT_REQUESTS = 'create_payment_requests';
    case MANAGE_PAYMENTS = 'manage_payments';
    case VIEW_FINANCIAL_REPORTS = 'view_financial_reports';
    
    // Communication
    case SEND_NOTIFICATIONS = 'send_notifications';
    case MODERATE_DISCUSSIONS = 'moderate_discussions';
    
    // Basic Access
    case VIEW_GROUP = 'view_group';
    case JOIN_SESSIONS = 'join_sessions';
    case LEAVE_GROUP = 'leave_group';
    case VIEW_BASIC_STATS = 'view_basic_stats';

    /**
     * Get permission label in Vietnamese
     */
    public function label(): string
    {
        return match($this) {
            self::EDIT_GROUP => 'Chỉnh sửa nhóm',
            self::DELETE_GROUP => 'Xóa nhóm',
            self::CHANGE_GROUP_SETTINGS => 'Thay đổi cài đặt nhóm',
            self::UPLOAD_GROUP_AVATAR => 'Tải ảnh đại diện nhóm',
            
            self::ADD_MEMBERS => 'Thêm thành viên',
            self::REMOVE_MEMBERS => 'Xóa thành viên',
            self::CHANGE_MEMBER_ROLES => 'Thay đổi vai trò thành viên',
            self::VIEW_MEMBER_DETAILS => 'Xem thông tin thành viên',
            self::APPROVE_MEMBER_REQUESTS => 'Phê duyệt yêu cầu tham gia',
            
            self::CREATE_SESSIONS => 'Tạo phiên tập/chơi',
            self::EDIT_SESSIONS => 'Chỉnh sửa phiên tập/chơi',
            self::DELETE_SESSIONS => 'Xóa phiên tập/chơi',
            self::MANAGE_ATTENDANCE => 'Quản lý điểm danh',
            
            self::CREATE_PAYMENT_REQUESTS => 'Tạo yêu cầu thanh toán',
            self::MANAGE_PAYMENTS => 'Quản lý thanh toán',
            self::VIEW_FINANCIAL_REPORTS => 'Xem báo cáo tài chính',
            
            self::SEND_NOTIFICATIONS => 'Gửi thông báo',
            self::MODERATE_DISCUSSIONS => 'Kiểm duyệt thảo luận',
            
            self::VIEW_GROUP => 'Xem nhóm',
            self::JOIN_SESSIONS => 'Tham gia phiên tập/chơi',
            self::LEAVE_GROUP => 'Rời khỏi nhóm',
            self::VIEW_BASIC_STATS => 'Xem thống kê cơ bản',
        };
    }

    /**
     * Get permission description
     */
    public function description(): string
    {
        return match($this) {
            self::EDIT_GROUP => 'Có thể chỉnh sửa thông tin cơ bản của nhóm',
            self::DELETE_GROUP => 'Có thể xóa nhóm (chỉ trưởng nhóm)',
            self::CHANGE_GROUP_SETTINGS => 'Có thể thay đổi cài đặt nhóm',
            self::UPLOAD_GROUP_AVATAR => 'Có thể tải lên và thay đổi ảnh đại diện nhóm',
            
            self::ADD_MEMBERS => 'Có thể mời và thêm thành viên mới',
            self::REMOVE_MEMBERS => 'Có thể loại bỏ thành viên khỏi nhóm',
            self::CHANGE_MEMBER_ROLES => 'Có thể thay đổi vai trò của thành viên',
            self::VIEW_MEMBER_DETAILS => 'Có thể xem thông tin chi tiết của thành viên',
            self::APPROVE_MEMBER_REQUESTS => 'Có thể phê duyệt yêu cầu tham gia nhóm',
            
            self::CREATE_SESSIONS => 'Có thể tạo phiên tập/chơi mới',
            self::EDIT_SESSIONS => 'Có thể chỉnh sửa thông tin phiên tập/chơi',
            self::DELETE_SESSIONS => 'Có thể xóa phiên tập/chơi',
            self::MANAGE_ATTENDANCE => 'Có thể quản lý điểm danh của thành viên',
            
            self::CREATE_PAYMENT_REQUESTS => 'Có thể tạo yêu cầu đóng phí',
            self::MANAGE_PAYMENTS => 'Có thể quản lý và xác nhận thanh toán',
            self::VIEW_FINANCIAL_REPORTS => 'Có thể xem báo cáo tài chính nhóm',
            
            self::SEND_NOTIFICATIONS => 'Có thể gửi thông báo cho thành viên',
            self::MODERATE_DISCUSSIONS => 'Có thể kiểm duyệt và quản lý thảo luận',
            
            self::VIEW_GROUP => 'Có thể xem thông tin nhóm',
            self::JOIN_SESSIONS => 'Có thể tham gia vào các phiên tập/chơi',
            self::LEAVE_GROUP => 'Có thể rời khỏi nhóm',
            self::VIEW_BASIC_STATS => 'Có thể xem thống kê cơ bản',
        };
    }
}